File: core/src/mindustry/client/navigation/MinePath.java
package mindustry.client.navigation;

import arc.*;
import arc.math.geom.*;
import arc.struct.*;
import arc.util.*;
import mindustry.gen.*;
import mindustry.type.*;
import mindustry.world.*;
import mindustry.world.blocks.storage.*;

import static mindustry.Vars.*;

public class MinePath extends Path {
    Seq<Item> items = new Seq<>(16);
    static Interval timer = new Interval();
    int cap = Core.settings.getInt("minepathcap");
    Item lastItem = null; // Last item mined

    public MinePath() {
        items = player.team().data().mineItems;
    }

    public MinePath(Seq<Item> mineItems, int cap) {
        items = mineItems;
        this.cap = cap;
    }

    public MinePath (String args){
        for (String arg : args.split("\\s")) {
            arg = arg.toLowerCase();
            boolean added = false;
            for (Item item : content.items().select(indexer::hasOre)) {
                if (item.name.toLowerCase().equals(arg) || item.localizedName.toLowerCase().equals(arg)) {
                    items.add(item);
                    added = true;
                }
            }
            if (!added) { // Item not matched
                if (arg.equals("*") || arg.equals("all") || arg.equals("a")) {
                    items.addAll(content.items().select(indexer::hasOre)); // Add all items when the argument is "all" or similar
                } else if (Strings.parseInt(arg) >= 0) {
                    cap = Strings.parseInt(arg);
                } else {
                    player.sendMessage(Core.bundle.format("client.path.builder.invalid", arg));
                }
            }
        }
        if (items.isEmpty()) {
            if (cap == Core.settings.getInt("minepathcap")) player.sendMessage(Core.bundle.get("client.path.miner.allinvalid"));
            items = player.team().data().mineItems;
        } else {
            player.sendMessage(Core.bundle.format("client.path.miner.mining", items.toString(", "), cap == 0 ? "infinite" : cap));
        }
    }

    @Override
    public void setShow(boolean show) {

    }

    @Override
    public boolean getShow() {
        return false;
    }

    @Override
    public void follow() {
        CoreBlock.CoreBuild core = player.closestCore();
        if (core == null) return;
        Item item = items.min(i -> indexer.hasOre(i) && player.unit().canMine(i), i -> core.items.get(i));
        if (item == null) return;

        if (lastItem != null && lastItem != item && core.items.get(lastItem) - core.items.get(item) < 100) item = lastItem; // Scuffed, don't switch mining until theres a 100 item difference, prevents constant switching of mine target
        lastItem = item;

        if (cap < core.storageCapacity && core.items.get(item) >= core.storageCapacity || cap != 0 && core.items.get(item) > cap) {  // Auto switch to BuildPath when core is sufficiently full
            player.sendMessage(Strings.format("[accent]Automatically switching to BuildPath as the core has @ items (this number can be changed in settings).", cap == 0 ? core.storageCapacity : cap));
            Navigation.follow(new BuildPath(items, cap == 0 ? core.storageCapacity : cap));
        }

        if (player.unit().maxAccepted(item) <= 1) { // drop off
            if (player.within(core, itemTransferRange - tilesize * 10) && timer.get(30)) {
                Call.transferInventory(player, core);
            } else {
                if (player.unit().type.canBoost) player.boosting = true;
                if (Core.settings.getBool("pathnav") && clientThread.taskQueue.size() == 0 && !player.within(core, itemTransferRange - tilesize * 15))
                    clientThread.taskQueue.post(() -> waypoints.set(Seq.with(Navigation.navigator.navigate(v1.set(player.x, player.y), v2.set(core.x, core.y), Navigation.obstacles.toArray(new TurretPathfindingEntity[0]))).filter(wp -> wp.dst(core) > itemTransferRange - tilesize * 15)));
                else waypoint.set(core.x, core.y, itemTransferRange - tilesize * 15, itemTransferRange - tilesize * 15);
            }

        } else { // mine
            Tile tile = indexer.findClosestOre(player.x, player.y, item);
            player.unit().mineTile = tile;
            if (tile == null) return;

            player.boosting = player.unit().type.canBoost && !player.within(tile, tilesize * 2); // FINISHME: Distance based on formation radius rather than just moving super close
            if (Core.settings.getBool("pathnav") && clientThread.taskQueue.size() == 0 && !player.within(tile, tilesize * 3))
                clientThread.taskQueue.post(() -> waypoints.set(Seq.with(Navigation.navigator.navigate(v1.set(player.x, player.y), v2.set(tile.getX(), tile.getY()), Navigation.obstacles.toArray(new TurretPathfindingEntity[0]))).filter(wp -> wp.dst(player) > tilesize)));
            else waypoint.set(tile.getX(), tile.getY(), tilesize, tilesize);
        }
        if (Core.settings.getBool("pathnav")) waypoints.follow();
        else waypoint.run(0);
    }

    @Override
    public void draw() {
        waypoints.draw();
    }

    @Override
    public float progress() {
        return 0;
    }

    @Override
    public void reset() {}

    @Override
    public Position next() {
        return null;
    }
}

File: core/src/mindustry/client/navigation/MinePath.kt
package mindustry.client.navigation

import arc.*
import arc.math.geom.*
import arc.struct.*
import arc.util.*
import mindustry.Vars.*
import mindustry.client.utils.*
import mindustry.content.*
import mindustry.game.*
import mindustry.gen.*
import mindustry.type.*
import mindustry.world.*
import mindustry.world.blocks.storage.*

class MinePath @JvmOverloads constructor(
    var items: Seq<Item> = Seq<Item>(),
    var cap: Int = Core.settings.getInt("minepathcap"),
    val newGame: Boolean = false,
    args: String = Core.settings.getString("defaultminepathargs")
) : Path() {

    private var lastItem: Item? = null // Last item mined
    private var timer = Interval()
    private var coreIdle = false
    private var bestItem: Item? = null
    var tile: Tile? = null

    init {
        val split = args.split("\\s".toRegex())
        if (items.isEmpty) {
            for (a in split) {
                if (a == "*" || a == "all" || a == "a") items.addAll(content.items().select(indexer::hasOre))
                else if (Strings.canParseInt(a)) cap = a.toInt().coerceAtLeast(0) // Specified cap, <= 0 results in infinite cap
                else content.items().find { a.equals(it.name, true) && indexer.hasOre(it) }?.apply(items::add) ?:
                player.sendMessage(Core.bundle.format("client.path.builder.invalid", a))
            }
        }

        if (items.isEmpty) {
            items.addAll(player.unit().type.mineItems)
            if (split.none { Strings.parseInt(it) > 0 }) player.sendMessage("client.path.miner.allinvalid".bundle())
        }
        else if (cap >= 0) {
            player.sendMessage(Core.bundle.format("client.path.miner.tobuild", items.joinToString(), if (cap == 0) "∞" else cap))
        } else {
            player.sendMessage(Core.bundle.format("client.path.miner.toidle", items.joinToString(), player.closestCore().storageCapacity))
        }
    }

    companion object {
        init {
            Events.on(EventType.WorldLoadEvent::class.java) {
                (Navigation.currentlyFollowing as? MinePath)?.lastItem = null // Reset on world load to prevent stupidity
            }
        }
    }

    override fun setShow(show: Boolean) = Unit
    override fun getShow() = false

    override fun follow() {
        val core = player.closestCore() ?: return
        val maxCap = if (cap <= 0) core.storageCapacity else core.storageCapacity.coerceAtMost(cap)
        bestItem = items.min({ indexer.hasOre(it) && player.unit().canMine(it) }) { core.items[it].toFloat() } ?: return
        if (lastItem != null && player.unit().canMine(lastItem) && indexer.hasOre(lastItem) && core.items[lastItem] - core.items[bestItem] < 100 && core.items[lastItem] < maxCap) bestItem = lastItem!! // Scuffed, don't switch mining until there's a 100 item difference, prevents constant switching of mine target
        lastItem = bestItem

        if (!newGame && core.items[bestItem] >= maxCap && cap >= 0) {  // Auto switch to BuildPath when core is sufficiently full
            coreIdle = false
            player.sendMessage(Core.bundle.format("client.path.miner.build", maxCap))
            Navigation.follow(BuildPath(items, cap))
        }

        // idle at core
        if (coreIdle) {
            if (!player.within(core, itemTransferRange - tilesize * 15)) goTo(core, itemTransferRange - tilesize * 15)

            if (player.unit().hasItem()) player.unit().clearItem() // clear items to prepare for MinePath resumption

            if (core.items[bestItem] < maxCap / 2) {
                player.sendMessage(Core.bundle.get("client.path.miner.resume"))
                coreIdle = false
            }

            return
        }

        // go to core and transfer items
        // No need to drop to core if within mineTransferRange
        if (player.unit().maxAccepted(bestItem) <= 1) {
            if (player.within(core, itemTransferRange - tilesize * 10) && timer[30f]) {
                player.unit().mineTile = null
                Call.transferInventory(player, core)

                // idle at core if cap < 0 (never switch to build path)
                if (core.items[bestItem] >= maxCap && cap < 0) {
                    player.sendMessage(Core.bundle.format("client.path.miner.idle", maxCap))
                    coreIdle = true
                }
            } else {
                if (player.unit().type.canBoost) player.boosting = true
                goTo(core, itemTransferRange - tilesize * 15)
            }

        // mine
        } else {
            tile = indexer.findClosestOre(player.unit(), bestItem) ?: return
            if (player.within(tile, player.unit().type.mineRange)) player.unit().mineTile = tile
            player.boosting = player.unit().type.canBoost && !player.within(tile, player.unit().type.mineRange)
            goTo(tile, player.unit().type.mineRange - tilesize * 2)
        }
    }

    override fun draw() {
        if ((waypoints.waypoints.lastOrNull()?.dst(player) ?: 0F) > tilesize * 3) waypoints.draw()
    }

    override fun progress() = 0F

    override fun reset() = Unit

    override operator fun next(): Position? = null

    // FINISHME: Unjank core tp on mix tech maps
//    override fun allowCore(core: CoreBlock.CoreBuild): Boolean {
//        val type = (core.block() as CoreBlock).unitType
//
//        if (tile == null) return false
//        val item: Item? =
//            if ((type.mineFloor && tile!!.block() == Blocks.air)) tile!!.drop()
//            else if (type.mineWalls) tile!!.wallDrop()
//            else return false
//
//        return item != null && type.mineTier >= item.hardness
}
